osu.Game.Rulesets.Osu.Tests/StackingTest.cs

```diff
@@ -18,6 +18,12 @@ namespace osu.Game.Rulesets.Osu.Tests
     [TestFixture]
     public class StackingTest
     {
+        [SetUp]
+        public void SetUp()
+        {
+            Decoder.RegisterDependencies(new AssemblyRulesetStore());
```

Comment:
The fact that tests are calling this method like that is just not sitting well with me. If this were called once ever in `OsuGameBase` I wouldn't have bat an eyelid, but given that this is a thing I can already imagine some sort of fiasco wherein tests are running in parallel and one test is overwriting the other's dependencies. But I don't know what the fix is - well, I have some ideas, and they all seem bad:

1. Get rid of all staticness in the decoders - tedious, long, hugely API-breaking presumably
2. Pass the dependencies to `.Decode()` every time - pure wrt side effects but tedious to work with, and I'm not even sure that's doable with the current structure of the base `Decoder`
3. Have some wild and dumb contraption like:
    - Keep `Decoder.RegisterDependencies()` as is but make it return an `IDisposable` / wait on a semaphore
    - As long as the `IDisposable` is not disposed / the semaphore is not signaled, nobody is allowed to call `RegisterDependencies()` again

Dunno. Do any of these sound even half acceptable? They sure don't to me, so I'll likely have to live with this as is...

Reply:
Another option is to automatically fall back to using an `AssemblyRulesetStore` if none is set manually. This would work in 99% of scenarios, and remove the test registration requirement.

Reply:
Out of all these I personally think the fallback to assembly is the best idea. I fear that making the beatmap resolve the ruleset is going to result in too many changes line-of-code-wise and a programming interface that is too brittle (need to remember to call some method before the beatmap is fully usable).

Reply:
It's just occurred to me that unless nunit is doing some extra process isolation that I don't know about, then even with the fallback idea there is a potential that some tests that currently manually do `Decoder.RegisterDependencies(new AssemblyRulesetStore());` may receive a `RealmRulesetStore` instead if they happen to run after a test that happened to register dependencies via `OsuGameBase` (because, well, static fields are static, as in CLR process-global)... Unsure if that is a huge issue.

Reply:
That’s correct. I don’t see it as a huge issue. Maybe if anything the abstract implementation should mention that returned `RulesetInfo` should be globally usable instances and nothing custom.